landingai/vision_pipeline.py

```diff
@@ -0,0 +1,338 @@
+        #     img = cv2.cvtColor(np.asarray(self.frames[0].image), cv2.COLOR_BGR2RGB)
+        # else:
+        #     img = cv2.cvtColor(np.asarray(self.frames[0].other_images[image_src]), cv2.COLOR_BGR2RGB)
+        # cv2.imshow("Landing AI - Press any key to exit", img)
+        # cv2.waitKey(0) # close window when a key press is detected
+        # cv2.waitKey(1)
+        # cv2.destroyWindow('image')
+        # for i in range (1,5):
+        #     cv2.waitKey(1)
+
+        return self
+
+    def apply(self, function: Callable[[Frame], Frame] = lambda f: f) -> "FrameSet":
+        """Apply a function to all frames
+
+        Parameters
+        ----------
+        function: lambda function that takes individual frames and returned an updated frame
+        """
+        for i in range(len(self.frames)):
+            self.frames[i] = function(self.frames[i])
+        return self
+
+    def filter(self, function: Callable[[Frame], bool] = lambda f: True) -> "FrameSet":
+        """Evaluate a function on every frame and keep or remove
+
+        Parameters
+        ----------
+        function : lambda function that gets invoked on every Frame. If it returns False, the Frame will be deleted
+        """
+        for i in reversed(
+            range(0, len(self.frames))
+        ):  # Traverse in reverse so we can delete
+            if not function(self.frames[i]):
+                self.frames.pop(i)
+        return self
+
+
+# openCV's default VideoCapture cannot drop frames so if the CPU is overloaded the stream will tart to lag behind realtime.
+# This class creates a treaded capture implementation that can stay up to date wit the stream and decodes frames only on demand
+class NetworkedCamera(BaseModel):
+    """The NetworkCamera class can connect to RTSP and other live video sources in order to grab frames. The main concern is to be able to consume frames at the source speed and drop them as needed to ensure the application allday gets the lastes frame"""
+
+    stream_url: str
+    motion_detection_threshold: int
+    capture_interval: float
+    previous_frame: Union[Frame, None] = None
+    _last_capture_time: datetime = PrivateAttr()
+    _cap: Any = PrivateAttr()  # cv2.VideoCapture
+    _FPS: int = PrivateAttr()
+    _lock: Any = PrivateAttr()  # threading.Lock
+    _t: Any = PrivateAttr()  # threading.Thread
+
+    def __init__(
+        self,
+        stream_url: str,
+        motion_detection_threshold: int = 0,
+        capture_interval: Union[float, None] = None,
+    ) -> None:
+        """
+        Parameters
+        ----------
+        stream_url : url to video source
+        motion_detection_threshold : If set to zero then motion detections is disabled. Any other value (0-100) will make the camera drop all images that don't have significant changes
+        capture_interval : If set to None, the NetworkedCamera will acquire images as fast as the source permits. Otherwise will grab new frames on the given interval
```

Comment:
nit: is the unit frame per second? it'd be nice to add a doc for it

Reply:
It is not FPS....I'll mention it on the doc